### PR TITLE
Remove BranchId from the workspace.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Can choose to review this commit by commit, or just the end form.

The BranchId concept has always been somewhat suspect.  It was a way of determining if the Solution instance you had was a Solution that had ever been teh .CurrentSolution of a workspace, or if you had a forked solution off of one of those solutions.  Needing to know if you were one of these has never been clear as to why that's important.  Roslyn operates in snapshots, and attempts to do so in a purely functional fashion.  What does it matter if you were on a .CurrentSolution, or a fork of that, or a fork of that fork?

At one point we definitely used this when it came to persisting data.  We didn't want analysis about a fork of a solution to overwrite data in our persistence service corresponding to the primary solution.  However, that code went away ages ago when:

1. we stopped storing diagnostic information in the persistence layer.
2. we ensured that all persisted data is stored with a checksum.

At that point, there was no concern about overwriting data with something invalid.

This PR gets rid of hte BranchId concept entirely and fixes up any remaining stragglers to get off of it.

### CODE REVIEW

File: src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs [hunk 1]
```diff
@@ -130,9 +130,6 @@ private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
         private async ValueTask SynchronizePrimaryWorkspaceAsync(CancellationToken cancellationToken)
         {
             var solution = _workspace.CurrentSolution;
-            if (solution.BranchId != _workspace.PrimaryBranchId)
-                return;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs:134]: totally unnecessary.  we are always accessing _workspace.CurrentSolution above.  So this solution was always the primary-branch solution.

File: src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs [hunk 2]
```diff
@@ -44,20 +45,18 @@ public Factory()
...
         /// <summary>
-        /// Cached compile time solution corresponding to the <see cref="Workspace.PrimaryBranchId"/>
+        /// Cached compile-time solution corresponding to an existing design-time solution.
         /// </summary>
-        private (int DesignTimeSolutionVersion, BranchId DesignTimeSolutionBranch, Solution CompileTimeSolution)? _primaryBranchCompileTimeCache;
+#if NETCOREAPP
+        private readonly ConditionalWeakTable<Solution, Solution> _designTimeToCompileTimeSolution = new();
+#else
+        private ConditionalWeakTable<Solution, Solution> _designTimeToCompileTimeSolution = new();
+#endif
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs:57]: .net core has `ConditionalWeakTable.Clear()`.  Standard does not.  So we simulate it by overwriting the field.

File: src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs [hunk 3]
```diff
@@ -44,20 +45,18 @@ public Factory()
...
 
-        /// <summary>
-        /// Cached compile time solution for a forked branch.  This is used primarily by LSP cases where
-        /// we fork the workspace solution and request diagnostics for the forked solution.
-        /// </summary>
-        private (int DesignTimeSolutionVersion, BranchId DesignTimeSolutionBranch, Solution CompileTimeSolution)? _forkedBranchCompileTimeCache;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs:60]: @dibarbet i believe this has the semantics from before.  Previously we keyed basd on Version+BranchId.  But these change every time you get a new solution.  So the only time these woudl match would be if you passed in teh exact same (reference-equals) Solution instance from before.  So we can get the same behavior by using a CWT keyed off Solution instances.

File: src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs [hunk 4]
```diff
@@ -57,8 +57,7 @@ internal class DiagnosticComputer
             _analysisKind = analysisKind;
             _analyzerInfoCache = analyzerInfoCache;
 
-            // We only track performance from primary branch. All forked branch we don't care such as preview.
-            _performanceTracker = project.IsFromPrimaryBranch() ? project.Solution.Services.GetService<IPerformanceTrackerService>() : null;
+            _performanceTracker = project.Solution.Services.GetRequiredService<IPerformanceTrackerService>();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs:60]: i don't see the value that in this.  Why would it not matter what the performance was getting diagnostics in a preview workspace?

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 5]
```diff
@@ -1202,16 +1195,6 @@ internal virtual bool TryApplyChanges(Solution newSolution, IProgressTracker pro
...
 
-                // make sure that newSolution is a branch of the current solution
-
-                // the given solution must be a branched one.
-                // otherwise, there should be no change to apply.
-                if (oldSolution.BranchId == newSolution.BranchId)
-                {
-                    CheckNoChanges(oldSolution, newSolution);
-                    return true;
-                }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:1213]: also not certain what teh value here is.  If someone asks us to apply changes, we can simply walk through and figure out waht the changes were.  This was a fast-bail-out for a case that seems like the unlikely one (e.g. when does someone ask to apply the same solution over the exisitng solution)?

File: src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs [hunk 6]
```diff
@@ -67,12 +66,15 @@ public CompileTimeSolutionProvider(Workspace workspace)
                 {
                     lock (_gate)
                     {
-                        _primaryBranchCompileTimeCache = null;
-                        _forkedBranchCompileTimeCache = null;
+#if NETCOREAPP
+                        _designTimeToCompileTimeSolution.Clear();
+#else
+                        _designTimeToCompileTimeSolution = new();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs:72]: note: doing the above seems unnecessary.  as a CWT, they should naturally release their resources when they are no longer referenced anywhere.

**Comment by CyrusNajmabadi** [comment]
@dibarbet @jasonmalinowski ptal.

File: src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs (hunk 3 above)
Comment (dibarbet) [reply src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs:60]: The idea here being that as soon as no one is using the solution we can get rid of the design time solution.  Makes sense to me.

**Review by dibarbet — APPROVED** [review]

**Pull request merged**